### DISCUSSION TITLE
Adding custom rules to my own project

### DISCUSSION THREAD

**Discussion by vorobeij** [cqfn/diKTat#747] [Q&A] [answered]
Hello,

I am wandering, can I add custom rules in my project without adding them to diktat repository?
I can do it with pure ktlint, but dikat has an amazing test framework. Can I add packages diktat-common and diktat-test-framework as gradle dependency to create my own rules?

Something like this (pseudocode):
app's `build.gradle`
```
apply diktat
apply ktlint
```
module's `build.gradle`
```
use ktlint rules from lint-rules module
```

**Project structure**
app
-module1
-module2
-lint-rules
--config.yaml
--src with my rules and tests for them

**Comment by orchestr7** [comment]
1) Of course you can use it as dependencies, our test-framework and common module are separate packages.  See: https://mvnrepository.com/artifact/org.cqfn.diktat/diktat-test-framework

2) Also - keep in mind that you can simply write your own ruleset (as diktat) and pass it to ktlint together with diktat ruleset. Ktlint supports multiple rulesets. We have even added special functionality to disable standard ktlint ruleset. See: https://github.com/pinterest/ktlint/pull/977
You will just need to run `ktlint --disabled_rules=standard -R yourRuleSet.jar, diktat.jar`  (or do the same using the ktlint gradle plugin)

**Comment by petertrr** [comment, marked as answer]
Hi, @SJOwl !
Yes, you can use `diktat-test-framework` as a dependency for testing your rules. It's a pretty generic framework that doesn't even depend on ktlint, so you'll need something like [our FixTestBase](https://github.com/cqfn/diKTat/blob/master/diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt): basically, there are two things:
* `TestComparatorUnit` is set up to run ktlint's `format` method
* `TestComparatorUnit.compareFilesFromResources` is called to run and compare output

Alternatively, you can include your rules in diktat itself: diktat uses the same rules format as ktlint, so providing a jar with your rules as a classpath dependency for `diktat-gradle-plugin` should be enough. To do so you can include it in `diktat` configuration that is created when applying diktat plugin, like that (I'm more familiar with gradle kotlin DSL than groovy, so there might be some errors):
```groovy
apply 'org.cqfn.diktat.diktat-gradle-plugin'

configurations.getByName("diktat").dependencies.add("your lint-rules")
```

**Reply by lyharrietbui** [reply]
Is there a way for me to write my own rule for diktat while diktat is configured within spotless? @petertrr

**Reply by orchestr7** [reply]
Yes, it is very easy - create your inspection, propose it to diktat with a PR - and after the release it will be automatically added to spotless.

**Reply by lyharrietbui** [reply]
@akuleshov7  Thank you for the answer. Is it not possible to do it without creating PR for diktat? I am trying to have a rule that reinforces braces for oneline if-else statement. I am not sure what the chance for the PR to be approved is

**Reply by orchestr7** [reply]
Strange - we should have such inspection. We do not reinforce braces?

**Reply by orchestr7** [reply]
Ah, single line. Single line is a ternary operator - it is not good to have braces in that. 
But! you can create a configuration for diktat for your use-case, it won't be the default in diktat, but you will be able to enable it

**Reply by orchestr7** [reply]
SO I suppose you can propose it as a configurable part of Braces rule

**Reply by lyharrietbui** [reply]
it says it enforce braces except for one-line if-else statement [here](https://github.com/analysis-dev/diktat/blob/master/diktat-analysis.yml#:~:text=%23%20Checks%20that%20braces%20are%20used%20in%20if%2C%20else%2C%20when%2C%20for%2C%20do%2C%20and%20while%20statements.%20Exception%3A%20single%20line%20if%20statement)

Does this mean that the following cases can have no braces? `val test = if (true) 3 else 4`

**Reply by lyharrietbui** [reply]
@akuleshov7 I see! How often do you have a new release?

**Reply by orchestr7** [reply]
> it says it enforce braces except for one-line if-else statement [here](https://github.com/analysis-dev/diktat/blob/master/diktat-analysis.yml#:~:text=%23%20Checks%20that%20braces%20are%20used%20in%20if%2C%20else%2C%20when%2C%20for%2C%20do%2C%20and%20while%20statements.%20Exception%3A%20single%20line%20if%20statement)
> 
> Does this mean that the following cases can have no braces? `val test = if (true) 3 else 4`

yes, this case won't have braces. As it is a ternary operator in Kotlin. But if you would like to have braces - you can propose a configuration to `diktat-analysis.yml` for braces rule

**Reply by orchestr7** [reply]
> @akuleshov7 I see! How often do you have a new release?

we plan to release next week. But we can wait for you patch

**Reply by lyharrietbui** [reply]
@akuleshov7  I am trying to create a PR! 
After forking and cloning and run `$ mvn clean install`
I got this error
```
-------------------------------------------------------------------------------
Test set: org.cqfn.diktat.plugin.maven.DiktatMavenPluginIntegrationTest
-------------------------------------------------------------------------------
Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 9.989 s <<< FAILURE! - in org.cqfn.diktat.plugin.maven.DiktatMavenPluginIntegrationTest
org.cqfn.diktat.plugin.maven.DiktatMavenPluginIntegrationTest.diktatFix(MavenExecutionResult)  Time elapsed: 4.258 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
	at diktat.maven.plugin@1.0.4-SNAPSHOT/org.cqfn.diktat.plugin.maven.DiktatMavenPluginIntegrationTest.diktatFix(DiktatMavenPluginIntegrationTest.kt:50)
```
Is this expected or am I missing some dependencies?

**Reply by orchestr7** [reply]
Which Java are you using? And which OS?

**Reply by lyharrietbui** [reply]
@akuleshov7 I am using
```
openjdk 11.0.9 2020-10-20 LTS
OpenJDK Runtime Environment Corretto-11.0.9.11.1 (build 11.0.9+11-LTS)
OpenJDK 64-Bit Server VM Corretto-11.0.9.11.1 (build 11.0.9+11-LTS, mixed mode)
```
 and `macOS Monterey version 12.3`

**Reply by lyharrietbui** [reply]
I changed the java version to `j8` and it seems to be fixing the issue

**Reply by orchestr7** [reply]
> I changed the java version to `j8` and it seems to be fixing the issue

this is what I was checking exactly in this right moment :)

**Reply by orchestr7** [reply]
> I changed the java version to `j8` and it seems to be fixing the issue

by the way - successful plugin builds are not necessary for your inspection

**Reply by lyharrietbui** [reply]
@akuleshov7 I see! The `diktat-maven-plugin` is successful but others are failing with `j8`
```
[INFO] diktat-maven-plugin ................................ SUCCESS [ 37.437 s]
[INFO] diktat-gradle-plugin ............................... FAILURE [07:35 min]
[INFO] org.cqfn.diktat.diktat-gradle-plugin.gradle.plugin . SKIPPED
```

**Reply by orchestr7** [reply]
> @akuleshov7 I see! The `diktat-maven-plugin` is successful but others are failing with `j8`
> 
> ```
> [INFO] diktat-maven-plugin ................................ SUCCESS [ 37.437 s]
> [INFO] diktat-gradle-plugin ............................... FAILURE [07:35 min]
> [INFO] org.cqfn.diktat.diktat-gradle-plugin.gradle.plugin . SKIPPED
> ```

yes, it is a problem with Jacoco report on MacOS, skip this project and that's it. It does not affect you at all :) 

We are building plugins on Linux in CI/CD.
`mvn -pl '!:diktat-gradle-plugin' clean install`

**Reply by lyharrietbui** [reply]
@akuleshov7 sound good. Thank you. Do you know a command  to fix the violation that is mentioned to be fixable by `diktat` configured within `spotless`? At the moment if I do if I can do `./gradlew :spotlessApply`, the build fail but the violation is not fixed automatically. I was trying to test the `NO_BRACES_IN_CONDITIONALS_AND_LOOPS` with these lines of code
```
        val test = if (true)
            println("hello")
            println("hi")
        else
            println("hi")
            println("ho")
```

**Reply by orchestr7** [reply]
BTW, you can try diktat online here: https://ktlint-demo.herokuapp.com
It will show how fixing capability in diktat works. Not every inspection has auto fixer

**Reply by orchestr7** [reply]
``` 
       val test = if (true)
            println("hello")
            println("hi")
        else
            println("hi")
            println("ho")
```

Try it with diktat directly. You can write a test, see examples at: `BracesRuleFixTest`

**Reply by orchestr7** [reply]
Also - the code you are providing is simply not compiling. There are no static analysers in the world that fix compilation issues :) This is why it is not fixed and you have exceptions

**Reply by lyharrietbui** [reply]
@akuleshov7 That's right. My bad. When I do `./gradlew :spotlessApply`, it actually fix the violation it is supposed to fix. I just need to fix a different error first that couldn't be fixed automatically

**Reply by lyharrietbui** [reply]
@akuleshov7 After reviewing the use case with my team, we think that the current multi if-else feature is sufficient for the time being. I am still interested in having a PR up but I am not sure when it would be yet but I don't think your release should get hung up by the PR!

**Reply by orchestr7** [reply]
> @akuleshov7 After reviewing the use case with my team, we think that the current multi if-else feature is sufficient for the time being. I am still interested in having a PR up but I am not sure when it would be yet but I don't think your release should get hung up by the PR!

sure, np

**Reply by lyharrietbui** [reply]
@akuleshov7 In the `dicktat-analysis.yml` file, there is a version of kotlin specified `kotlinVersion: 1.5` in `DIKTAT_COMMON`. is `DIKTAT_COMMON` also a rule that can be enabled and disabled?. Can this be interpolated? Does this break if a service is using a lower or higher version of kotlin? I was testing with a service using higher kotlin version and it seems to work. I want to make sure if we need to match the version number.
